### PR TITLE
[FEAT] Args syscall filter

### PR DESCRIPTION
**Pull request by NDStrahilevitz**
## Initial Checklist

- [x] There is an issue describing the need for this PR (#2177).
- [x] Git log contains summary of the change.
- [x] Git log contains motivation and context of the change.
- [ ] If part of an EPIC, PR git log contains EPIC number.
- [ ] If part of an EPIC, PR was added to EPIC description.

## Description (git log)

commit 2182908a7ff0fd8c8153624e02f7c591a5249e08 (HEAD -> args_syscall_filter, origin/args_syscall_filter)
```
    filters: parse syscall argument filters by name
    
    This allows generic usage of syscall event names in relevant argument
    filters.
    
    Without this change, syscall ids may differ between architectures.
```

commit 7b2a68f3e06e434663db2c526f2c3be091fd7725
```
    filters: add generic Filter method to interface
    
    1. Add the generic Filter method, this takes any value to the filter.
    In practice, the external method must confirm the type, then pass to an
    internal type-safe method.
    
    2. Add generic type constraints to int and uint filters. This enables
    later down the line to avoid correct type casting for arguments.
```

## Type of change

- [ ] Bug fix (non-breaking change fixing an issue, preferable).
- [ ] Quick fix (minor non-breaking change requiring no issue, use with care)
- [ ] Code refactor (code improvement and/or code removal)
- [x] New feature (non-breaking change adding functionality).
- [ ] Breaking change (cause existing functionality not to work as expected).

## How Has This Been Tested?

1) Run `tracee-ebpf -t security_file_open.syscall=execve -o option:detect-syscall -t e=security_file_open`
Confirm that events are outputted only with "execve" syscalls and that most events are filtered:
`Stats: {EventCount:46 EventsFiltered:1322 NetEvCount:0 ErrorCount:0 LostEvCount:0 LostWrCount:0 LostNtCount:0}`

2) Run `tracee-ebpf -t task_rename.syscall=execnotve`
Confirm that tracee fails running.

## Final Checklist:

Pick "Bug Fix" or "Feature", delete the other and mark appropriate checks.

- [x] I have made corresponding changes to the documentation.
- [x] My code follows the style guidelines (C and Go) of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented all functions/methods created explaining what they do.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] My changes generate no new warnings.
- [ ] I have added tests that prove my fix, or feature, is effective.
- [x] New and existing unit tests pass locally with my changes.
- [x] Any dependent changes have been merged and published before.

## Git Log Checklist:

My commits logs have:

- [x] Subject starts with "subsystem|file: description".
- [x] Do not end the subject line with a period.
- [x] Limit the subject line to 50 characters.
- [x] Separate subject from body with a blank line.
- [x] Use the imperative mood in the subject line.
- [x] Wrap the body at 72 characters.
- [x] Use the body to explain what and why instead of how.

### CODE REVIEW

**Comment by NDStrahilevitz** [comment]
@rafaeldtinoco I think the check-vet fails here because of a golang version mismatch on your runner machines, could you confirm if they support go.18?

**Review by geyslan — COMMENTED** [review]
LGTM

**Comment by rafaeldtinoco** [comment]
Starting to review this now.

File: pkg/filters/args.go [hunk 1]
```diff
@@ -92,13 +95,33 @@ func (filter *ArgFilter) Parse(filterName string, operatorAndValues string, even
 		return InvalidEventArgument(argName)
 	}
 
+	var err error
+	if argName == "syscall" {
+		err = filter.parseSyscallFilter(id, operatorAndValues)
+	} else {
+		err = filter.parseFilter(id, argName, operatorAndValues, func() Filter {
+			return NewStringFilter()
```
Comment (rafaeldtinoco) [inline pkg/filters/args.go:103]: Do you expect to have other types here ? is that the reason why you want a constructor pointer ?

**Review by rafaeldtinoco — APPROVED** [review]
@NDStrahilevitz sorry for taking so long for reviewing this but I went back up to when you changed all the filters (needed to catch up with your code for this and further PRs).

So, it all looks good to me (your already done filter work is very good). I left some small questions.

Also, it feels to me that "generics" would be better used if we joined Int and Uint filters into a single object using "generics" (supporting s8, s16, s32, s64, u8, u16, u32, u64), rather than using it to differentiate among 32bit and 64bit versions only.  

Of course I'm not proposing this to be done in this PR, so I'm approving this one but I'm a bit skeptical in the generics usage (at least for this PR, maybe you're expecting to be use in a different way in further PRs).

Cheers!

File: pkg/filters/uint.go [hunk 2]
```diff
@@ -24,16 +25,18 @@ type UIntFilter struct {
...
-func NewUInt32Filter() *UIntFilter {
-	return newUIntFilter(true)
+func NewUInt32Filter() *UIntFilter[uint32] {
+	return newUIntFilter[uint32](true)
 }
 
-func newUIntFilter(is32Bit bool) *UIntFilter {
-	return &UIntFilter{
+func newUIntFilter[T constraints.Unsigned](is32Bit bool) *UIntFilter[T] {
+	return &UIntFilter[T]{
```
[the remark this answers is not in the document]
Comment (NDStrahilevitz) [reply pkg/filters/uint.go:39]: Basically if you take a look at the external `Filter` method it does the following:
`filterable, ok := val.(T)` this is basically the magic of the generic type parameter, as long as we construct the correct Filter for the argument type, this type cast should always succeed.

I believe, as you said, that we can merge the Int and Uint filters as well, I haven't thought of that before, nice catch. Let's do that in another PR.

File: pkg/filters/args.go (hunk 1 above)
Comment (NDStrahilevitz) [reply pkg/filters/args.go:103]: In the future when we better map out argument types (i'm prototyping that in my spare time), we can easily map an argument to a filter constructor.

**Pull request merged**